### ISSUE TITLE
Wavesurfer instances crashes Safari iOS 

### ISSUE THREAD

**Issue by Stianlars1** [katspaugh/wavesurfer.js#2590]
Hello! This will be my first post here, and I'm currently in a big hurry, but I´ll try to explain fast: 

So I'm building this app in react for visualizing audio waveforms. The user is allowed to upload songs to firebase, and in my homepage.js I run a component that returns the title, artist, download URL from firebase and cover-photo. 

In my homepage I return a div for each element I get from firebase, and I initialize the wave surfer component I made for each element I get from firebase (instances of a song). 

To prevent memory leaks I've decided to slice my array, and have only 10 listings at once showing, so the user need to click a button to get the next 10 songs form the array, and the old ones will get overlooked. 

I currently have 42 songs in firebase, and when I click my way to page 4 (song number: 40) the website crashes, or actually it reloads, and get stuck in a loop where it´s just loading, loading, and returning to page 1. 

page 1 - 2 and 3 works like a charm, but page 4 and 5 doesn't.. 

It works perfect in Chrome desktop, safari desktop and Firefox desktop. but safari and chrome on iOS (iPhone) just fucks up.. 

Im not sure what to add here, but I´ll show the code so you get an understanding. 

-- I've been toggling between "WebAudio", "MediaElement" and the others without anyone of the solving the problem.

After some reloads and refreshes safari just displays this problem: 

"A problem repeatedly occured"

Please help! tnx

code::

this is from my AudioVisualizer.js component:

```
  useEffect(()=>{
          if (audioRef.current){
                  audioTrackRef.current = wavesurfer.create({
                  container: audioRef.current,
                  progressColor: "#13AEA2",
                  waveColor: "red",
                  cursorColor: "OrangeRed",
                  preload: true,
                  backend: "WebAudio", // originally = "MediaElement"
                  barWidth: 2,
                  barHeight: 1, // the height of the wave
                  fillParent: true,
                  hideScrollbar: true,
                  responsive: true,
                  
              });

              audioTrackRef.current.load(props.audio);

        
        }
 }, [])   
```

My homepage.js where I create the instance:

```
                    <div className='homepage_container'>
                        <h1 className='homeTitleSongs'>Songs</h1>
                        <button className='logOutButton' onClick={handleLogout}>Logout</button>
                        <div className='changeSongList'>
                            <div className='changeSongListButtons'>
                            <button className='changeSongListButton' onClick={updateMapSlicePrevSongs}>Prev songlist</button>
                            <button className='changeSongListButton' onClick={updateMapSliceNextSongs}>Next songlist</button>
                            </div>
                            <h2>{atThisPage} / {Math.ceil((songs.length / 10))}</h2>
                        </div>




                        {  
                                songs.length && songs.slice(mapSliceCount1, mapSliceCount2).map((data) => {

                                return (


                                    <article key={data.id} className='card'> 
                                        <div className='card_content'>
                                            <img style={{width : "160px", height : "160px"}} className='card_image' src={data.image} alt=""/>

                                            <div className='song_info'>
                                            <h2>{data.title}</h2>
                                            <h4>{data.artist}</h4>
                                            </div>


                                            <div className='audio_wrapper'>
                                            <AudioVisualizer audio={data.audio}/>
                                            </div>

                                        </div>
                                    

                                    </article>
                                )
                            }) 

                        }
                    </div>
```

Inside of Homepage I also have this function that will slice the map function and show the next 10 songs from the list. I may have a thought of that if I want to go to page 4, when I click the next 10 songs it will load first all 10 songs on page 2, then all 10 songs on page 3 before getting to page 4. maybe there's a problem here? idk.. 

function for going to next songs are here:

   ```
 function updateMapSliceNextSongs() {
        let countSongsInList = songs.length;
        let countedSongInList_converted = (Math.floor((countSongsInList / 10)) * 10);
        let countedSongInList_converted2 = ((Math.floor((countSongsInList / 10)) * 10) + 10);

        console.log(countedSongInList_converted2);
        if (mapSliceCount1 === countedSongInList_converted && mapSliceCount2 === countedSongInList_converted2) {
            return
        } else {
            setMapSliceCount1(mapSliceCount1 => mapSliceCount1 += 10);
            setMapSliceCount2(mapSliceCount2 => mapSliceCount2 += 10);
            setAtThisPage(atThisPage => atThisPage += 1);
            console.log(mapSliceCount1)
        }

    }
    function updateMapSlicePrevSongs() {
        if (mapSliceCount1 === 0 && mapSliceCount2 === 10){
            return
        } else {
            setMapSliceCount1(mapSliceCount1 => mapSliceCount1 -= 10);
            setMapSliceCount2(mapSliceCount2 => mapSliceCount2 -= 10);
            setAtThisPage(atThisPage => atThisPage -= 1);

        }

    }

```

**Comment by thijstriemstra** [comment]
> It works perfect in Chrome desktop, safari desktop and Firefox desktop. but safari and chrome on iOS (iPhone) just fucks up..

Please add browser verion and operating system details for each browser you tested.

**Comment by sundayz** [comment]
Use https://github.com/bbc/audiowaveform to create the waveform data as either json or images and just have 1 player on the page. If you load 10+ audio players it's going to use too much memory and the page will crash, because the mp3 file needs to be loaded and decoded in memory to show the waveform.

**Comment by Stianlars1** [comment]
I solved the problem by storing the audio download links in an array, and then slice the array when mapping through it for display. I sliced it (0, 10) which results in no problems loading the 10 songs. 

Then I added 2 buttons to navigate to next -> 10 songs, and 10 <- prev songs. which seems to work. 

Another problem I stumbled across was that when I clicked the next -> button for lets say page 4 // slice(40, 50) it resulted in a crash, maybe because it will need to load all songs going from 0-10, then 10-20, then 20-30 etc.. 

Solved this issue by storing the page the user was on in localStorage, so if the site refreshes or crashes, it will load that page it crashed on.
